### PR TITLE
Alerting: Clamp Loki ASH range query to configured max_query_length

### PR DESCRIPTION
**Pull request by rwwiv**
**What is this feature?**

This PR clamps the time range of Loki queries using the state historian Loki client to the configured `max_query_length` value in order to avoid request failures.

**Why do we need this feature?**

If Grafana is set up to query Loki for alert state history, annotation queries that span a time longer than `max_query_length` (721h = 30d1h by default) fail on the Loki side, which means state history does not appear in panels.

**Who is this feature for?**

Users of dashboard annotations who have configured Grafana to save/fetch alert state history from a Loki database.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ngalert/state/historian/loki_http.go [hunk 1]
```diff
@@ -297,3 +286,36 @@ type QueryRes struct {
...
+			c.log.Error("Error response from Loki with an empty body", "status", res.StatusCode)
+		}
+		return nil, fmt.Errorf("received a non-200 response from loki, status: %d", res.StatusCode)
+	}
+
+	return data, nil
+}
+
+// ClampRange ensures that the time range is within the configured maximum query length.
+func ClampRange(start, end, maxTimeRange int64) (newStart int64, newEnd int64) {
```
Comment (alexweav) [inline pkg/services/ngalert/state/historian/loki_http.go:313]: nit: could we add a unit test in this package for this? Right now the logic is here but the tests live elsewhere.

File: pkg/services/ngalert/state/historian/loki_http.go [hunk 2]
```diff
@@ -261,23 +258,15 @@ func (c *HttpLokiClient) RangeQuery(ctx context.Context, logQL string, start, en
 	if err != nil {
 		return QueryRes{}, fmt.Errorf("error executing request: %w", err)
 	}
-
 	defer func() {
-		_ = res.Body.Close()
+		if err := res.Body.Close(); err != nil {
+			c.log.Warn("Failed to close response body", "err", err)
+		}
```
Comment (alexweav) [inline pkg/services/ngalert/state/historian/loki_http.go:264]: :+1:

**Review by alexweav — APPROVED** [review]
LGTM, just a nit :smile:

File: pkg/services/ngalert/state/historian/loki_http.go (hunk 1 above)
Comment (rwwiv) [reply pkg/services/ngalert/state/historian/loki_http.go:313]: Ah yeah good call 👍

**Pull request merged**